crates/joystick/src/mapping.rs

```diff
@@ -0,0 +1,128 @@
+use crate::common::{Button, JoystickInterface};
+use crate::joystick::Joystick;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+pub struct GenericJoystick<KeyType: PartialEq + Eq + Hash> {
+    joystick: Joystick,
+    mappings: HashMap<KeyType, Button>,
+}
+
+impl<KeyType: PartialEq + Eq + Hash> GenericJoystick<KeyType> {
+    pub fn new(mappings: HashMap<KeyType, Button>) -> Self {
+        Self {
+            joystick: Joystick::default(),
+            mappings,
+        }
+    }
+
+    pub fn release_all(&mut self) {
+        self.joystick.release_all();
+    }
+    pub fn press(&mut self, key: &KeyType) {
+        if let Some(button) = self.mappings.get(key) {
+            self.joystick.press(button);
+        } // Note: could panic! here on None?
+    }
+    pub fn release(&mut self, key: &KeyType) {
+        if let Some(button) = self.mappings.get(key) {
+            self.joystick.release(button);
+        } // Note: could panic! here on None?
+    }
+    // [x, y], where the values range from -1 to 1
+    pub fn set_ljoy(&mut self, dir: [f32; 2]) {
+        self.joystick.set_ljoy(dir);
+    }
+
+    pub fn set_rjoy(&mut self, dir: [f32; 2]) {
+        self.joystick.set_rjoy(dir);
+    }
+}
+
+// To visually run these tests, use:
+// `cargo test -- --test-threads 1`
+// then focus https://hardwaretester.com/gamepad
+#[cfg(test)]
+mod tests {
+    // Used in the test
+    use std::thread::sleep;
+    use std::time::Duration;
+
+    // Include the library
+    use crate::prelude::*;
+
+    // This test doubles as an example on how to use the joystick library with custom actions
+    #[derive(PartialEq, Eq, Hash)]
+    enum Action {
+        Confirm,
+        Cancel,
+        Fire,
+        Jump,
+        Aim,
+    }
+
+    #[test]
+    fn test_mappings() -> std::io::Result<()> {
+        sleep(Duration::from_millis(2000));
+        let mut gamepad = GenericJoystick::new(HashMap::from([
+            (Action::Confirm, Button::A),
+            (Action::Cancel, Button::B),
+            (Action::Fire, Button::X),
+            (Action::Jump, Button::Y),
+            (Action::Aim, Button::RT(255)),
```

Comment:
Wasn't this intended to be -1 to 1?

Reply:
This is the trigger buttons, not the joysticks. I made a direct mapping in Button::LT/RT to the internal type u8. I suppose we could abstract it, but it'd be 0.0 (not pressed) to 1.0 (fully pressed) in that case.